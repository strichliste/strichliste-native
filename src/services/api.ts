const API_URL = "https://demo.strichliste.org/api/";

export async function fetchJson(
  endpoint: string,
  options: RequestInit = {}
): Promise<any> {
  return fetch(`${API_URL}${endpoint}`, options).then(async res => res.json());
}

export async function post(endpoint: string, params: any): Promise<any> {
  const options: RequestInit = {
    mode: "cors",
    method: "POST",
    body: JSON.stringify(params),
    headers: { "Content-Type": "application/json; charset=utf-8" }
  };
  return fetch(`${API_URL}${endpoint}`, options).then(async res => res.json());
}

export async function get(endpoint: string): Promise<any> {
  return fetch(`${API_URL}${endpoint}`).then(async res => res.json());
}

export async function restDelete(endpoint: string): Promise<any> {
  return fetch(`${API_URL}${endpoint}`, {
    method: "delete"
  }).then(async res => res.json());
}
