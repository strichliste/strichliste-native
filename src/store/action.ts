import {
  ArticleActions,
  ErrorActions,
  LoaderActions,
  SearchActions,
  SettingsActions,
  TransactionActions,
  UserActions,
} from './reducers';

export type Action =
  | ArticleActions
  | ErrorActions
  | LoaderActions
  | SettingsActions
  | SearchActions
  | TransactionActions
  | UserActions;
