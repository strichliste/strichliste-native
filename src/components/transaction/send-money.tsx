import React, { useState } from "react";
import { User } from "../../store/reducers";
import { View } from "react-native";

export const SendMoney = () => {
  const [user, setUser] = useState<User | null>(null);

  return <View />;
};
