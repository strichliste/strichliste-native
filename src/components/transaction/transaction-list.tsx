import React from "react";
import { Text, View } from "react-native";
import Icon from "react-native-vector-icons/FontAwesome";

import { User, Article, getTransaction } from "../../store/reducers";
import { store } from "../../store";
import { Currency } from "../ui/text";
import { getTheme } from "../ui/theme";

const theme = getTheme();

interface ListItemProps {
  user?: User;
  article?: Article;
  isSender: boolean;
  comment?: string;
  gotoUser(user: User): void;
}

const ListItemDescription = ({
  user,
  article,
  isSender,
  comment,
  gotoUser
}: ListItemProps) => {
  const composedComment = (user && comment ? ":" : "") + (comment || "");
  return (
    <Text numberOfLines={1} ellipsizeMode="tail">
      {user && (
        <Text onPress={() => gotoUser(user)}>
          {isSender ? <>&#8592;</> : <>&#8594;</>} {user.name}
        </Text>
      )}
      {article && (
        <>
          <Icon name="shopping-cart" />
          <Text>{article.name}</Text>
        </>
      )}
      <Text>{composedComment}</Text>
    </Text>
  );
};

interface Props {
  id: number | string;
  gotoUser(user: User): void;
}

export function TransactionListItem({
  id,
  gotoUser
}: Props): JSX.Element | null {
  const transaction = getTransaction(store.getState(), Number(id));

  if (!transaction) {
    return null;
  }

  return (
    <View
      style={{
        borderBottomColor: theme.border,
        borderBottomWidth: 1,
        paddingBottom: 8,
        paddingTop: 8
      }}
    >
      <View
        style={{
          flexDirection: "row",
          justifyContent: "space-between"
        }}
      >
        <Currency value={transaction.amount} />
        <Text numberOfLines={1} ellipsizeMode="tail">
          {transaction.created}
        </Text>
      </View>
      <ListItemDescription
        gotoUser={gotoUser}
        article={transaction.article}
        isSender={!!transaction.sender}
        comment={transaction.comment}
        user={transaction.sender || transaction.recipient}
      />
    </View>
  );
}
