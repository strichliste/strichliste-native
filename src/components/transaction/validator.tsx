import { Boundary } from "../../store/reducers";
import { store } from "../../store";
import React from "react";

interface TransactionArguments {
  accountBoundary: Boundary;
  paymentBoundary: Boundary;
  isDeposit: boolean;
  balance: number | boolean;
  value: number;
}

interface CheckValidProps {
  accountBoundaryValue: number | boolean;
  paymentBoundaryValue: number | boolean;
  balance: number | boolean;
  value: number;
}

function checkDepositIsValid({
  accountBoundaryValue,
  paymentBoundaryValue,
  balance,
  value
}: CheckValidProps): boolean {
  if (
    typeof paymentBoundaryValue === "number" &&
    value > paymentBoundaryValue
  ) {
    return false;
  }

  if (
    typeof accountBoundaryValue === "boolean" ||
    typeof balance === "boolean"
  ) {
    return true;
  }
  return value + balance < accountBoundaryValue;
}

function checkDispenseIsValid({
  accountBoundaryValue,
  paymentBoundaryValue,
  balance,
  value
}: CheckValidProps): boolean {
  if (
    typeof paymentBoundaryValue === "number" &&
    value > paymentBoundaryValue * -1
  ) {
    return false;
  }

  if (
    typeof accountBoundaryValue === "boolean" ||
    typeof balance === "boolean"
  ) {
    return true;
  }
  return balance - value > accountBoundaryValue;
}

export const isTransactionValid = ({
  accountBoundary,
  paymentBoundary,
  isDeposit,
  balance,
  value
}: TransactionArguments): boolean => {
  if (value === 0) {
    return false;
  }
  if (isDeposit) {
    return checkDepositIsValid({
      accountBoundaryValue: accountBoundary.upper,
      paymentBoundaryValue: paymentBoundary.upper,
      value,
      balance
    });
  } else {
    return checkDispenseIsValid({
      accountBoundaryValue: accountBoundary.lower,
      paymentBoundaryValue: paymentBoundary.lower,
      value,
      balance
    });
  }
};

export function useTransactionValidator(
  value: number,
  userId: string,
  isDeposit: boolean = true
): boolean {
  const state = store.getState();
  const settings = state.settings;
  const user = state.user[userId];
  const balance = user ? user.balance : 0;

  return isTransactionValid({
    value,
    balance,
    isDeposit,
    accountBoundary: settings.account.boundary,
    paymentBoundary: settings.payment.boundary
  });
}

interface UserToUserValidatorProps {
  userId: string;
  targetUserId: string;
  value: number;
  render(isValid: boolean): JSX.Element;
}

export function UserToUserValidator(
  props: UserToUserValidatorProps
): JSX.Element | null {
  const userHasTheMoney = useTransactionValidator(
    props.value,
    props.userId,
    false
  );
  const receiverCanAcceptTheMoney = useTransactionValidator(
    props.value,
    props.targetUserId,
    true
  );
  return <>{props.render(userHasTheMoney && receiverCanAcceptTheMoney)} </>;
}
