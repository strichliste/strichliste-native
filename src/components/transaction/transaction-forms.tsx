import React from "react";
import { View, Text } from "react-native";

import { Card } from "../ui/card";
import { store } from "../../store";
import { startCreatingTransaction } from "../../store/reducers";
import { Button } from "../ui/form";
import { useTransactionValidator } from "./validator";
import { formatCurrency } from "../ui/text";

export interface PaymentButtonListProps {
  steps: number[];
  userId: string;
  isDeposit: boolean;
}

interface Props {
  userId: string;
}

export function Payment(props: Props): JSX.Element {
  const payment = store.getState().settings.payment;
  return (
    <Card>
      {payment.deposit.enabled && (
        <PaymentButtonList
          isDeposit={true}
          steps={payment.deposit.steps}
          userId={props.userId}
        />
      )}
      <View>
        <Text>Test</Text>
      </View>
      {payment.dispense.enabled && (
        <PaymentButtonList
          isDeposit={false}
          steps={payment.dispense.steps}
          userId={props.userId}
        />
      )}
    </Card>
  );
}

const chunk = (arr: any[], chunkSize: number) => {
  return arr
    .slice(0, ((arr.length + chunkSize - 1) / chunkSize) | 0)
    .map(function(c, i) {
      return arr.slice(chunkSize * i, chunkSize * i + chunkSize);
    });
};

export function PaymentButtonList(props: PaymentButtonListProps): JSX.Element {
  const multiplier = props.isDeposit ? 1 : -1;
  const stepGroups = chunk(props.steps, 3);
  return (
    <View>
      {stepGroups.map((steps, index) => (
        <View
          key={index}
          style={{
            flexDirection: "row",
            marginBottom: 8
          }}
        >
          {steps.map(step => (
            <View style={{ minWidth: 75, marginRight: 8, flex: 1 }} key={step}>
              <TransactionButton
                isDeposit={props.isDeposit}
                userId={props.userId}
                value={step * multiplier}
              />
            </View>
          ))}
        </View>
      ))}
    </View>
  );
}

interface TransactionButtonProps {
  userId: string;
  value: number;
  isDeposit?: boolean;
}

export function TransactionButton(props: TransactionButtonProps): JSX.Element {
  const isValid = useTransactionValidator(
    props.value,
    props.userId,
    props.isDeposit
  );

  return (
    <Button
      onPress={() =>
        startCreatingTransaction(store.dispatch, props.userId, {
          amount: props.value
        })
      }
      isGreen={props.isDeposit}
      isRed={!props.isDeposit}
      isDisabled={!isValid}
      title={formatCurrency(props.value)}
    />
  );
}
