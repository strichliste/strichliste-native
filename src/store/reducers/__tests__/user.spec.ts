/* eslint-disable @typescript-eslint/no-explicit-any */

import { DeepPartial } from 'redux';
import { user } from '..';
import { get, post } from '../../../services/api';
import { getMockStore } from '../../../spec-configs/mock-store';
import { Action } from '../../action';
import { transactionsLoaded } from '../transaction';
import {
  UserActionTypes,
  getUser,
  getUserArray,
  getUserTransactionsArray,
  startCreatingUser,
  startLoadingUserDetails,
  startLoadingUsers,
  startUpdateUser,
  userDetailsLoaded,
} from '../user';

jest.mock('../../../services/api', () => ({
  get: jest.fn(),
  post: jest.fn(),
}));

describe('user reducer', () => {
  let action: DeepPartial<Action>;

  describe('with non matching action', () => {
    it('returns initial state', () => {
      const initialState = {};
      expect(user(undefined, {} as any)).toEqual(initialState);
    });
  });

  describe('with a UsersLoaded action', () => {
    const usersResponse = {
      users: [
        {
          id: 1,
          name: 'schinken',
          isActive: true,
          email: null,
          balance: 12330,
          created: '2018-08-18 16:18:40',
        },
      ],
    };
    beforeEach(() => {
      action = {
        type: UserActionTypes.UsersLoaded,
        payload: usersResponse,
      };
    });

    it('creates a new user entry in the store', () => {
      const expectedResult = {
        1: {
          id: 1,
          name: 'schinken',
          isActive: true,
          email: null,
          balance: 12330,
          created: '2018-08-18 16:18:40',
        },
      };
      const result = user(undefined, action as Action);
      expect(result).toEqual(expectedResult);
    });
  });

  describe('with a UserDetailsLoaded action', () => {
    it('adds the user to the state', () => {
      action = userDetailsLoaded({
        id: 1,
        name: 'schinken',
        isActive: true,
        balance: 12330,
        created: '2018-08-18 16:18:40',
      });
      const expectedResult = {
        1: {
          id: 1,
          name: 'schinken',
          isActive: true,
          balance: 12330,
          created: '2018-08-18 16:18:40',
        },
      };
      const result = user(undefined, action as Action);
      expect(result).toEqual(expectedResult);
    });
  });

  describe('with a TransactionsLoaded action', () => {
    describe('and a matching user', () => {
      it('updates transactions for this user', () => {
        action = transactionsLoaded([{ id: 1, user: { id: 'user1' } }] as any);
        const result = user(
          { user1: { id: 'user1', transactions: { 2: 2 } } } as any,
          action as Action
        );
        const expectedResult = {
          user1: { id: 'user1', transactions: { 1: 1, 2: 2 } },
        };
        expect(result).toEqual(expectedResult);
      });
    });
    describe('without a matching user', () => {
      it('sets transaction and user of the response', () => {
        action = transactionsLoaded([{ id: 1, user: { id: 'user1' } }] as any);
        const result = user(undefined, action as Action);
        const expectedResult = {
          user1: { id: 'user1', transactions: { 1: 1 } },
        };
        expect(result).toEqual(expectedResult);
      });
    });
    describe('with an empty transaction', () => {
      it('wont throw an exception', () => {
        action = transactionsLoaded([] as any);
        const result = user(undefined, action as Action);
        expect(result).toEqual(result);
      });
    });
  });
});

describe('action creators', () => {
  describe('startLoadingUsers', () => {
    it('triggers usersLoaded on success', async () => {
      (get as any).mockImplementationOnce(() => Promise.resolve([{ id: 1 }]));

      const store = getMockStore();
      await startLoadingUsers(store.dispatch, true);
      expect(get).toHaveBeenCalledWith('user?deleted=false&active=true');
      expect(store.getActions()).toMatchSnapshot();
    });
  });

  describe('startLoadingUserDetails', () => {
    it('triggers userDetailsLoaded on success', async () => {
      (get as any).mockImplementationOnce(() =>
        Promise.resolve({ user: [{ id: 1 }] })
      );

      const store = getMockStore();
      await startLoadingUserDetails(store.dispatch, 1);
      expect(get).toHaveBeenCalledWith('user/1');
      expect(store.getActions()).toEqual([
        userDetailsLoaded([{ id: 1 }] as any),
      ]);
    });
  });

  describe('startCreatingUser', () => {
    it('triggers userDetailsLoaded on success', async () => {
      (post as any).mockImplementationOnce(() =>
        Promise.resolve({ user: [{ id: 1 }] })
      );
      const store = getMockStore();
      await startCreatingUser(store.dispatch, 'test');
      expect(post).toHaveBeenCalledWith('user', { name: 'test' });
      expect(store.getActions()).toMatchSnapshot();
    });
  });

  describe('startUpdateUser', () => {
    it('triggers userDetailsLoaded on success', async () => {
      (post as any).mockImplementationOnce(() =>
        Promise.resolve({ user: [{ id: 1 }] })
      );
      const store = getMockStore();
      await startUpdateUser(store.dispatch, 1, {
        name: 'test',
        isDisabled: true,
      });
      expect(post).toHaveBeenCalledWith('user/1', {
        name: 'test',
        isDisabled: true,
      });
      expect(store.getActions()).toMatchSnapshot();
    });
  });
});

describe('selectors', () => {
  describe('getUserArray', () => {
    it('returns the current user map as an array', () => {
      expect(
        getUserArray({
          user: { 1: { id: 1, name: 'a' }, 2: { id: 2, name: 'b' } },
        } as any)
      ).toEqual([{ id: 1, name: 'a' }, { id: 2, name: 'b' }]);
    });
  });
  describe('getUser', () => {
    it('selects a user by id', () => {
      expect(
        getUser(
          {
            user: { 1: { id: 1 }, 2: { id: 2 } },
          } as any,
          2
        )
      ).toEqual({ id: 2 });
    });
  });
  describe('getUserTransactionsArray', () => {
    describe('with no user', () => {
      it('returns an empty array', () => {
        expect(
          getUserTransactionsArray(
            { user: { 1: { id: 1 }, 2: { id: 2 } } } as any,
            4
          )
        ).toEqual([]);
      });
    });
    describe('with no transactions', () => {
      it('returns an empty array', () => {
        expect(
          getUserTransactionsArray(
            { user: { 1: { id: 1 }, 2: { id: 2 } } } as any,
            2
          )
        ).toEqual([]);
      });
    });
    describe('with transactions set', () => {
      it('selects all user transactions as an array sorted desc by id', () => {
        expect(
          getUserTransactionsArray(
            {
              user: {
                1: { id: 1 },
                2: { id: 2, transactions: { 2: 2, 3: 3, 1: 1, 8: 8 } },
              },
            } as any,
            2
          )
        ).toEqual([8, 3, 2, 1]);
      });
    });
  });
});
