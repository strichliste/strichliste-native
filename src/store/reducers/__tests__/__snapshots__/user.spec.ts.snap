// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`action creators startCreatingUser triggers userDetailsLoaded on success 1`] = `
Array [
  Object {
    "payload": Object {
      "GLOBAL_LOADER": true,
    },
    "type": "SET_LOADER",
  },
  Object {
    "payload": Object {
      "GLOBAL_ERROR": "",
    },
    "type": "SET_ERROR",
  },
  Object {
    "payload": Object {
      "GLOBAL_LOADER": false,
    },
    "type": "SET_LOADER",
  },
  Object {
    "payload": Array [
      Object {
        "id": 1,
      },
    ],
    "type": "[USER] Loaded_Details",
  },
]
`;

exports[`action creators startLoadingUsers triggers usersLoaded on success 1`] = `
Array [
  Object {
    "payload": Object {
      "GLOBAL_LOADER": true,
    },
    "type": "SET_LOADER",
  },
  Object {
    "payload": Object {
      "GLOBAL_ERROR": "",
    },
    "type": "SET_ERROR",
  },
  Object {
    "payload": Object {
      "GLOBAL_LOADER": false,
    },
    "type": "SET_LOADER",
  },
  Object {
    "payload": Array [
      Object {
        "id": 1,
      },
    ],
    "type": "[USER] Loaded",
  },
]
`;

exports[`action creators startUpdateUser triggers userDetailsLoaded on success 1`] = `
Array [
  Object {
    "payload": Object {
      "GLOBAL_LOADER": true,
    },
    "type": "SET_LOADER",
  },
  Object {
    "payload": Object {
      "GLOBAL_ERROR": "",
    },
    "type": "SET_ERROR",
  },
  Object {
    "payload": Object {
      "GLOBAL_LOADER": false,
    },
    "type": "SET_LOADER",
  },
  Object {
    "payload": Array [
      Object {
        "id": 1,
      },
    ],
    "type": "[USER] Loaded_Details",
  },
]
`;
